primitives/state-machine/src/testing.rs

```diff
@@ -172,6 +172,12 @@ impl<H: Hasher, N: ChangesTrieBlockNumber> TestExternalities<H, N>
 		self.backend.update(transaction)
 	}
 
+	/// Commit all pending changes to the underlying backend.
+	pub fn commit_all(&mut self) {
+		self.backend = self.as_backend();
+		self.overlay.clear();
+	}
```

Comment:
Why can't we just call `self.ext().commit()`?

Reply:
I still see this `commit` function as a hack that I would like to remove at some point. 

Do you don't like the `clear()` call here?

Reply:
I was just wondering why the duplication of the the logic that writes the changes into the backend. Meaning `self.as_backend()` vs `self.overlay.drain_storage_changes()`-

Reply:
I am seconding a bit this, but I am fine without it (the PR already remove some duplication).
'clear' did frightened me a bit at first (get rid of changes trie records so can be confusing), but using 'Default instead to reset layer requires copying some fields and I am not sure it is any better, maybe it could be restricted to the crate, or be rename to 'reset'.

Reply:
@athei & @cheme please check my latest commit.